VideoCommon: rework anamorphic widescreen heuristic

ligfx opened the pull request:
Some widescreen hacks (see below) properly force anamorphic output, but don't make the last projection in a frame 16:9, so Dolphin doesn't display it correctly.

This changes the heuristic code to assume a frame is anamorphic based on the total number of vertex flushes in 4:3 and 16:9 projections that frame. It also adds a bit of "aspect ratio inertia" by making it harder to switch aspect ratios, which takes care of aspect ratio flickering that some games / widescreen hacks would be susceptible with the new logic.
    
I've tested this on SSX Tricky's native anamorphic support, Tom Clancy's Splinter Cell (it stayed in 4:3 the whole time), and on the following widescreen hacks for which the heuristic doesn't currently work:

Paper Mario: The Thousand-Year Door (Gecko widescreen code from Nintendont)
```
C202F310 00000003
3DC08042 3DE03FD8
91EEF6D8 4E800020
60000000 00000000
04199598 4E800020
C200F500 00000004
3DE08082 3DC0402B
61CE12A2 91CFA1BC
60000000 387D015C
60000000 00000000
C200F508 00000004
3DE08082 3DC04063
61CEE8D3 91CFA1BC
60000000 7FC3F378
60000000 00000000
```

The Simpsons: Hit & Run (AR widescreen code from the wiki)
```
04004600 C002A604
04004604 C09F0014
04004608 FC002040
0400460C 4082000C
04004610 C002A608
04004614 EC630032
04004618 48220508
04041A5C 38600001
04224344 C002A60C
04224B1C 4BDDFAE4
044786B0 3FAAAAAB
04479F28 3FA33333
```

JMC47 commented:
@phire can you review this?  You wrote the heuristic.

phire commented:
I feel like this is a bad idea. I have seen games which mess up and manage to trigger a false 16:9 positive every-so-often even with just checking the last projection (Tom Clancy's Splinter Cell).

The current heuristic is not ideal, but I think something like "If the majority of projections are 16:9" would work better.

ligfx commented:
@phire I'll work on logic for "If the majority of projections are 16:9". Do you happen to remember where in Splinter Cell it does that? If you do, I can go through and test that this won't give false positives.

phire commented:
The splinter cell issue happened semi-reliably when playing through the tutorial mission.

ligfx commented:
@phire I've updated the heuristic to count vertex flushes rather than projection view changes (which fixes practically every issues I saw in Splinter Cell), and added a bit of inertia to make it less likely for the aspect ratio to flicker (which fixes the rest of the issues with Splinter Cell). The widescreen hacks I was using now work as expected as well (i.e. mostly widescreen, occasional menus and GUI elements in standard).

When you find some time, would you mind taking another look at this?

phire commented:
Ok, now you have added inertia, I have to wonder how it copes with the widescreen AR code for wind waker.

The interesting thing about that AR code is only the 3d scenes are 16:9, with all the 2d menus sticking with 4:3. With the old version of the heuristic, dolphin would instantly flip between 4:3 and 16:9 as you entered/exited the pause menu.

ligfx commented:
@phire using the widescreen Gecko code from the Dolphin wiki, the Wind Waker menu/item screen now stays at 16:9! Screenshots attached:

Current:
<img width="400" alt="screen shot 2017-04-09 at 3 53 36 pm" src="https://cloud.githubusercontent.com/assets/594093/24841624/50f35b40-1d3d-11e7-907c-ecf29c99c186.png">

With this patch:
<img width="400" alt="screen shot 2017-04-09 at 3 56 20 pm" src="https://cloud.githubusercontent.com/assets/594093/24841625/55dcc844-1d3d-11e7-84b3-d69ac2d876aa.png">

phire commented:
Not only does this do a better job at detecting AR/Gecko widescreen codes, it prevents flickering in at least two 4:3 only games (Spinter Cell and PacMan World Rally).

Merging.

phire commented:
Fixes [Issue 9895](https://bugs.dolphin-emu.org/issues/9895)
